From c5f88ded782c8896e04d7d4ce36bbb7157bd87c5 Mon Sep 17 00:00:00 2001
From: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Date: Thu, 5 Oct 2023 22:00:07 +0300
Subject: [PATCH] xen/domain: Save PMU counters to mdcr_el2

According to arm documentation
"When EL2 is implemented and enabled for the current
 Security state, reads of this field from EL1
 and EL0 return the value of MDCR_EL2.HPMN."

To be able to use PMU counters from guest domain, the
counters should be read from pmcr_el0 register and saved
to mdcr_el2.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Co-authored-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
---
 xen/arch/arm/domain.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/xen/arch/arm/domain.c b/xen/arch/arm/domain.c
index 68d2f7fcbd..646dc485f1 100644
--- a/xen/arch/arm/domain.c
+++ b/xen/arch/arm/domain.c
@@ -561,6 +561,11 @@ void free_vcpu_struct(struct vcpu *v)
     free_xenheap_pages(v, get_order_from_bytes(sizeof(*v)));
 }
 
+uint32_t get_pmu_counters(void)
+{
+    return (READ_SYSREG(PMCR_EL0) >> 11) & 0x1F;
+}
+
 int arch_vcpu_create(struct vcpu *v)
 {
     int rc = 0;
@@ -592,6 +597,8 @@ int arch_vcpu_create(struct vcpu *v)
     v->arch.mdcr_el2 = HDCR_TDRA | HDCR_TDOSA | HDCR_TDA;
     if ( !(v->domain->options & XEN_DOMCTL_CDF_vpmu) )
         v->arch.mdcr_el2 |= HDCR_TPM | HDCR_TPMCR;
+    else
+        v->arch.mdcr_el2 |= get_pmu_counters();
 
     if ( (rc = vcpu_vgic_init(v)) != 0 )
         goto fail;
-- 
2.25.1

